pkg/gardenlet/controller/federatedseed/seed_control.go

```diff
@@ -0,0 +1,249 @@
+		for {
+			select {
+			case res := <-c.workerCh:
+				c.numberOfRunningWorkers += res
+				logger.Logger.Debugf("Current number of running Seed Federated controller workers is %d", c.numberOfRunningWorkers)
+			}
+		}
+	}()
+
+	logger.Logger.Infof("Starting federated controllers")
+	for i := 0; i < workers; i++ {
+		controllerutils.CreateWorker(ctx, c.seedQueue, "Seed", c.createReconcileSeedRequestFunc(ctx), &waitGroup, c.workerCh)
+	}
+
+	<-ctx.Done()
+
+	logger.Logger.Infof("Shutting down federated controllers")
+	c.federatedSeedControllerManager.ShutDownControllers()
+	c.seedQueue.ShutDown()
+	waitGroup.Wait()
+}
+
+func (c *Controller) createReconcileSeedRequestFunc(ctx context.Context) reconcile.Func {
+	return func(req reconcile.Request) (reconcile.Result, error) {
+		seed, err := c.seedLister.Get(req.Name)
+		if apierrors.IsNotFound(err) {
+			logger.Logger.Infof("Skipping - Seed %s was not found", req.Name)
+			return reconcile.Result{}, nil
+		}
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		if seed.DeletionTimestamp != nil {
+			c.federatedSeedControllerManager.removeController(seed.GetName())
+			return reconcile.Result{}, nil
+		}
+
+		condition := helper.GetCondition(seed.Status.Conditions, gardencorev1beta1.SeedBootstrapped)
+		if condition.Status != gardencorev1beta1.ConditionTrue {
+			return reconcile.Result{}, nil
+		}
+
+		if err := c.federatedSeedControllerManager.createControllers(ctx, seed.GetName(), c.k8sGardenClient, c.config, c.recorder); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+	}
+}
+
+type federatedSeedControllerManager struct {
+	seedShootStateSyncController map[string]*shootstate.SyncController
+	lock                         sync.RWMutex
+}
+
+func (f *federatedSeedControllerManager) createControllers(ctx context.Context, seedName string, k8sGardenClient kubernetes.Interface, config *config.GardenletConfiguration, recorder record.EventRecorder) error {
+	if _, found := f.seedShootStateSyncController[seedName]; found {
+		logger.Logger.Infof("Controllers are already started for seed %s", seedName)
+		return nil
+	}
+
+	seedLogger := logger.Logger.WithField("Seed", seedName)
+	k8sSeedClient, err := seedpkg.GetSeedClient(ctx, k8sGardenClient.Client(), config.SeedClientConnection.ClientConnectionConfiguration, config.SeedSelector == nil, seedName)
+	if err != nil {
+		return err
+	}
+
+	extensionsClient, err := extensionsclientset.NewForConfig(k8sSeedClient.RESTConfig())
+	if err != nil {
+		return err
+	}
+
+	extensionsInformers := extensionsinformers.NewSharedInformerFactory(extensionsClient, 0)
+	shootStateSyncController := shootstate.NewSyncController(ctx, k8sGardenClient, k8sSeedClient, extensionsClient, extensionsInformers, config, seedLogger, recorder)
+
+	logger.Logger.Infof("Run ShootStateSync controller for seed: %s", seedName)
+	if err := shootStateSyncController.Run(ctx, *config.Controllers.ShootStateSync.ConcurrentSyncs); err != nil {
```

Comment:
Just saw this:
We pass the same context to all new sync controllers we create and it is later used to create the workers.
If a sync controller has to be removed there is no way to actually stop the worker loop: `wait.Until(worker(queue, resourceType, reconciler), time.Second, ctx.Done())`
`controller.Stop()` doesn't close the context so the `wait.Until` will keep looping forever.

What we could do is:
1. Create a new child context here
2. Pass the child context to Run()
3. Save the child context's cancel func in the controller's entry in the controller map with `addController()`
4. Call cancel when we call `removeController()

I have another idea how to make it with calling defer cancel() right after creating the context but it's a bit more complicated and if we anyway refactor the workers it might not be necessary to do it now.